### DISCUSSION TITLE
Example of use cmnd for Modbus - MQTT publish format?

### DISCUSSION THREAD

**Discussion by martin8310** [arendst/Tasmota#17009] [General]
Hello,

Could you explain me which publish MQTT format I have to use for that :

works fine in console:
`ModBusSend {"deviceaddress": 1, "functioncode": 3, "startaddress": 0, "type":"uint32", "count":1}`

result:
`tele/test111_C91F6C/RESULT = {"ModbusReceived": "DeviceAddress":1,"FunctionCode":3,"StartAddress":0,"Length":9,"Count":1,"Values":[59]}}`

thank you in advance.

regards
Martin

**Comment by sfromis** [comment]
Publish to topic `cmnd/test111_C91F6C/ModBusSend` with payload `{"deviceaddress": 1, "functioncode": 3, "startaddress": 0, "type":"uint32", "count":1}`
Or you can publish to topic `cmnd/test111_C91F6C/Backlog` with payload `ModBusSend {"deviceaddress": 1, "functioncode": 3, "startaddress": 0, "type":"uint32", "count":1}`

How exactly to do that depends on your tool of choice.

**Comment by martin8310** [comment]
Thank you! It works!

**Comment by NickB1** [comment]
I'm having troubles with this.

Using the console with this command works fine:

`CMD: ModBusSend {"deviceaddress": 1, "functioncode": 16, "startaddress": 0, "type":"uint16", "count":9, "values":[65535,65535,65535,65535,0,0,0,0,1]}`
`MQT: stat/dimmer-gardenhouse/RESULT = {"ModbusSend":"Done"}`
`MQT: tele/dimmer-gardenhouse/RESULT = {"ModbusReceived":{"DeviceAddress":1,"FunctionCode":16,"StartAddress":0,"Length":8,"Count":9}}`

Publishing to `cmnd/dimmer-gardenhouse/ModBusSend` with `{"deviceaddress": 1, "functioncode": 16, "startaddress": 0, "type":"uint16", "count":9, "values":[65535,65535,65535,65535,0,0,0,0,1]}` gives `stat/dimmer-gardenhouse/RESULT = {"Command":"Error"}` in the console.

Any ideas? Do I have to compile Tamosta with the USE_MODBUS_BRIDGE define or should this work by default?
I'm using tasmota32.bin

**Reply by sfromis** [reply]
Since issuing the command in the console works, you have some error in how you handle MQTT. Are you sure that you split the JSON part out to the separate field for payload, not in the same field as the topic?
If you first activate `Weblog 3`, then the console should show details for what Tasmota received, including what command it could not recognize, but not splitting into the payload field would be a typical cause.

**Reply by barbudor** [reply]
As per the docs [Modbus Bridge](https://tasmota.github.io/docs/Modbus-Bridge/) is included in Tasmota32

Did you configured the device `dimmer-gardenhouse` with the Modbus TX & RX GPIO ?
ModbusBridge needs obviously to be connected to a Modbus with the appropriate adapter and the GPIO configured

**Reply by sfromis** [reply]
Since the ModBusSend command worked when issued from the console, it does seem that the feature is both included in the binary and also configured.

**Reply by NickB1** [reply]
When activating Weblog 3 I get the following when issuing the command
`CMD: Grp 0, Cmd 'MODBUSSEND', Idx 1, Len 119, Pld -99, Data '{deviceaddress: 1, functioncode: 16, startaddress: 0, type:uint16, count:9, values:[65535,65535,65535,65535,0,0,0,0,1]}'`

**Reply by NickB1** [reply]
Figured it out, made a silly mistake.
I used double quotation marks to enclose the command to mosquitto_pub but this didn't play well with the double quotation marks within the json string.
Using single quotation marks to enclose the command fixed the issue:
` '{"deviceaddress": 1, "functioncode": 16, "startaddress": 0, "type":"uint16", "count":9, "values":[0,0,0,0,0,0,0,0,1]}'`
